pkg/core/resources/apis/mesh/mesh_validator.go

```diff
@@ -63,33 +68,45 @@ func validateLoggingBackend(backend *mesh_proto.LoggingBackend) validators.Valid
 	if err := accesslog.ValidateFormat(backend.Format); err != nil {
 		verr.AddViolation("format", err.Error())
 	}
-	if file, ok := backend.GetType().(*mesh_proto.LoggingBackend_File_); ok {
-		verr.AddError("file", validateLoggingFile(file))
-	} else if tcp, ok := backend.GetType().(*mesh_proto.LoggingBackend_Tcp_); ok {
-		verr.AddError("tcp", validateLoggingTcp(tcp))
+	switch backend.GetType() {
+	case mesh_proto.LoggingFileType:
+		verr.AddError("config", validateLoggingFile(backend.Config))
+	case mesh_proto.LoggingTcpType:
+		verr.AddError("config", validateLoggingTcp(backend.Config))
+	default:
+		verr.AddViolation("type", fmt.Sprintf("unknown backend type. Available backends: %q, %q", mesh_proto.LoggingTcpType, mesh_proto.LoggingFileType))
 	}
 	return verr
 }
 
-func validateLoggingTcp(tcp *mesh_proto.LoggingBackend_Tcp_) validators.ValidationError {
+func validateLoggingTcp(cfgStr *structpb.Struct) validators.ValidationError {
 	var verr validators.ValidationError
-	if tcp.Tcp.Address == "" {
-		verr.AddViolation("address", "cannot be empty")
+	cfg := mesh_proto.TcpLoggingBackendConfig{}
+	if err := proto.ToTyped(cfgStr, &cfg); err != nil {
```

Comment:
```go
func validateLoggingTcp(cfgStr *structpb.Struct) validators.ValidationError {
	var verr validators.ValidationError
	cfg := mesh_proto.TcpLoggingBackendConfig{}
	if err := proto.ToTyped(cfgStr, &cfg); err != nil {
		verr.AddViolation("", fmt.Sprintf("could not parse config: %s", err.Error()))
		return verr
	} 

	if cfg.Address == "" {
		verr.AddViolation("address", "cannot be empty")
		return verr
	} 
	
	host, port, err := net.SplitHostPort(cfg.Address)
	if host == "" || port == "" || err != nil {
		verr.AddViolation("address", "has to be in format of HOST:PORT")
	}

	return verr
}
```
If I'm correct this is the same method but feels more readable to me because of less nesting. Please consider using

Reply:
The problem is that it's not the same. With your suggestion we validate regardless if the config is completely broken (not convertible to a type). I wanted to avoid that

Reply:
But if it is broken we just return `verr`:
```go
if err := proto.ToTyped(cfgStr, &cfg); err != nil {
	verr.AddViolation("", fmt.Sprintf("could not parse config: %s", err.Error()))
	return verr
} 
```

Reply:
ah, I see. Ok, let me change it then